crates/matrix-sdk/src/widget_api/messages/from_widget/actions.rs

```diff
@@ -23,3 +27,32 @@ pub struct SupportedVersions {
 pub enum ApiVersion {
     PreRelease,
 }
+
+// MSC1960
+
+#[derive(Serialize, Deserialize, Debug)]
+pub enum OpenIdState {
+    #[serde(rename = "allowed")]
+    Allowed,
+    #[serde(rename = "blocked")]
+    Blocked,
+    #[serde(rename = "request")]
+    PendingUserConfirmation,
+}
+#[derive(Serialize, Deserialize, Debug)]
+pub struct GetOpenIdResponse {
+    state: OpenIdState, //OpenIDRequestState;
+    original_request_id: String,
+    access_token: Option<String>,
+    expires_in: Option<i32>,
+    matrix_server_name: Option<String>,
+    token_type: Option<String>,
+}
+
+#[derive(Serialize, Deserialize, Debug)]
+struct SendToDeviceRequest {
+    #[serde(rename = "type")]
+    message_type: String,
+    encrypted: bool,
+    content: HashMap<String, HashMap<String, serde_json::Value>>,
```

Comment:
The usage of `serde_json::Value` gives some "might be wrong" vibes 🙂 Are we confident that we just pass this `Value` around without really parsing it?

Reply:
the actual content of the to device request can be very generic: the content field itself has a user key and inside a user key the device keys. But the actual message can be anything (including custom json that a widget developer might want to send)

Reply:
Thanks for the clarification! Should at least the keys be then somewhat specified? (some concrete type with a concrete name?)

Feel free to resolve upon reading :)

Reply:
I think on this level not anymore. The event consists of type, and content. The fields of the contend depend on the type but the type can be any string. (there are of course specced types that define a content but it is also possible to introduce any new type in a widget with arbitrary content. -> at this stage one cannot tell anything about the structure of content)